JL_STD* vs ios_std* cleanup

samoconnor opened the pull request:
Use jl_printf(), jl_puts() etc instead of ios_printf(), JL_PRINTF(), fprintf() etc for consistency.

Make jl_printf(JL_STD*, ...), jl_puts() safe in very early stages of initialisation.

Try to remove miscellaneous ios.c dependencies.
Remaining ios.c dependencies are:
- flisp/*
- iostream.jl (and supporting C functions in sys.c) -- uses ios.c to do file IO.
- dump.c, -- uses ios_mem() for RAM buffers and ios_file() to read/write files.

See detail in commit logs and discussion here:
https://groups.google.com/forum/?fromgroups=#!topic/julia-dev/UlfIRvX0_nw

tkelman commented:
This has timed out twice on win32 now, https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.1021/job/eku72d9iugsgkaym and https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.1023/job/633jwx6bgernyteq (I cancelled that second one after it was stuck for about 10 minutes), both at the very end of building the system image. I'm thinking that might not be a coincidence.

tkelman commented:
Yep, this branch freezes locally for me on win32. Also needs another rebase (slightly preferred to resolve conflicts via rebasing than merging)

samoconnor commented:
I've reset back to 584cd7d526f8b3ede9d05fd18703096c20ba94a8, I'll make sure win32 works, then rebase to resolve conflicts...
(I'm trying to get win32 build going on a WinXP VM).

tkelman commented:
> WinXP VM

If you have a choice to do something newer, your probability of getting a successful from-source compile will be higher. You can also save some dependency-building time by trying out the script I use on AppVeyor, `contrib/windows/msys_build.sh`.

samoconnor commented:
I'm afraid XP is the newest win32 I have a licence for. Last time I checked, people who wanted to be non bleeding edge still used XP.
My usual purpose for the XP VM is opening the odd .xls file that Numbers.app doesn't understand.
Perhaps I should get a licence for something newer.
What version is currently best for dev work?

tkelman commented:
I'd recommend any of win7, win8, or server 2012 (or 2012 R2). There are short-term evaluation licenses that you can get for free. If you're using the newest version of Packer and Vagrant, the steps from here: https://github.com/staticfloat/julia-vagrant/pull/2#issuecomment-58506560 were pretty painless

tkelman commented:
We've been trying to keep the binaries running on XP, but building on XP is another story

samoconnor commented:
Waiting for this now...
```
curl -O -L "https://www.modern.ie/vmdownload?platform=mac&virtPlatform=vmware&browserOS=IE11-Win8.1&parts=4&filename=VMBuild_20140402/VMware/IE11_Win8.1/Mac/IE11.Win8.1.For.MacVMWare.part{1.sfx,2.rar,3.rar,4.rar}"
```

From : https://www.modern.ie/en-us/virtualization-tools

I've never heard of or used Packer or Vagrant.

I'm hoping that once I have the windows 8.1 VM up and running, I can follow the steps here to build julia : https://github.com/JuliaLang/julia/blob/master/README.windows.md ...

tkelman commented:
> I've never heard of or used Packer or Vagrant.

No problem, they're just convenient automation wrappers around VirtualBox / VMWare. Not mandatory at all.

> I'm hoping that once I have the windows 8.1 VM up and running, I can follow the steps here to build julia : https://github.com/JuliaLang/julia/blob/master/README.windows.md ...

Yes, though building all the dependencies will take quite some time. The cygwin-cross version is a bit more actively tested by myself and I think most other active Windows Julia developers lately than MSYS2, but if one method works and not the other then I'd consider it a bug.

Using the AppVeyor script would allow the build to take ~15 minutes as opposed to a few hours, due to building all of the dependencies.

samoconnor commented:
OK, it sounds like the plan is:
 - https://cygwin.com/setup-x86_64.exe
 - + make, curl, p7zip, and mingw64-$ARCH-gcc-g++
 - contrib/windows/msys_build.sh

tkelman commented:
And git. Run the script as `ARCH=i686 contrib/windows/msys_build.sh` to get a 32-bit build (otherwise the script will guess based on `uname`). That works locally for me in a clean clone, so barring any other dependencies I didn't fully declare, it should hopefully work for you too!

samoconnor commented:
Any reason not to build ARCH=x86_64 ?

tkelman commented:
Actually I guess both were timing out, so good point. Keeping separate clones for `julia32` vs `julia64` is easier than switching between the two (the script doesn't do any cleanup, it makes a bit of a mess of the clone you run it in with downloaded files).

vtjnash commented:
windows 10 is also a free download right now from microsoft

samoconnor commented:
Thanks for your advice on the windows build Tony.
ARCH=i686 passes tests on my Win 8.1 VMware box now.
I haven't tried 64 bit because apparently the windows 8.1 image I have does not support 64 bit...

tkelman commented:
> Thanks for your advice on the windows build Tony.

No problem, glad it worked and you found the issue. Not many people are willing to fire up a VM to fix Windows issues, so thank you!

> I haven't tried 64 bit because apparently the windows 8.1 image I have does not support 64 bit...

D'oh. I'm checking this branch locally, it looks okay so far. It'll take AppVeyor a long time to get to it, I think the queue right now is the longest it's been yet. It's working on build 1066 right now, this PR is build 1080, and we have up to 1092 queued.

samoconnor commented:
Hi All,
Is there any interest in merging this PR?
If so I can do another pass of rebase/resolve-conflicts/squash in the next day or so.
After that my available time will be greatly reduced.

ViralBShah commented:
Please do so. The cleanup is nice to have and best to keep the PR ready to merge.

samoconnor committed:
Refactor to reduce dependancy on ios.c

Rationale:

    Unravel ios.c’s legacy role as debug io library for the runtime;
    vs current role as backend for iostream.jl.

Clean up historical debug printf functions:

    Replace calls to ios_printf(), ios_puts(), JL_PRINTF, JL_PUTS etc
    with jl_printf(), jl_puts()...

    Replace exit() with jl_exit() in repl.c, dump.c and init.c.

Make jl_printf() and jl_exit() safe in early initialisation.

Remove ios.c dependance from ast.c and builtins.c:

    src/flisp/flisp.c:

        New fl_load_system_image_str(char*, size_t)
        Hides detail of ios_static_buffer from ast.c
        ios_* no longer used in ast.c.

    src/builtins.c

        In jl_errorf(), replace ios_vprintf() with vasprintf()
        (following the precedent set by src/jl_uv.c:663).

Use free() not LLT_FREE() to clean up after vasprintf() in jl_uv.c:

    man vasprintf says “asprintf() and vasprintf() dynamically
                        allocate a new string with malloc(3)”

Refactor pairs of calls to jl_printf() & jl_exit(1) into calls to
jl_error[f](), which does the same thing during early init.

Dead code removal:

    jl_uv.c: jl_bufptr(), jl_ios_buf_base(), jl_putc(), jl_pututf8()
    sys.c: jl_ios_size()

New jl_safe_printf().

jl_safe_printf() uses a statically allocated 1000 byte buffer to
vnsprintf() a message then calls write(2) to send it to stderr.

jl_safe_printf() is intended to be used in places where there is the
potential for out of memory, corrupted stack, re-entrancy etc.

The current jl_safe_printf() implementation is not 100% safe in all
situations, but it should be pretty good most of the time, especially
if the formatting is limited to %d.

The safety of jl_safe_printf() could be improved later, but at least
for now the places were safety might matter are identified by its use.

jl_safe_printf() is used in:

 - sigdie_handler()
 - sigint_handler()
 - _exception_handler()
 - catch_exception_raise()
 - gdblookup()

Thanks to @vtjnash for assistance and feedback with these changes.

Dead code cleanup & jl_write() simplificaiton.

Clean up dead code:
    stream.jl : write!() — My apologies if this is used somewhere, I
could’t find any doc or callers.
    builtins.c : jl_print_symbol(), jl_print_int64()
    ccall.cpp : if(0 …) jl_puts()
    jl_uv.c : jl_puts

Replace 2 remaining calls to jl_puts() with calls to jl_printf() for
consistency.
    builtins.c : jl_error()
    repl.c : true_main()

Move jl_uv_writecb() implementation from stream.jl to jl_uv.c.
This callback is only used by jl_write(),
which is only called from jl_vprintf() in runtime C code.
This helps to de-tangle fs.jl and stream.jl from the runtime's printf().

Use jl_safe_printf() to attempt error message output in jl_uv_writecb().
If this callback is printing an error, there was a problem writing to
JL_STDOUT or JL_STDERR, so its best to try the simpelest means possible
to get the error out.

update doc/devdocs/stdio.rst to remove reference to:
    jl_write(), jl_putc(), jl_puts().

Simplify handling of STDOUT and STDERR early in initialisation:
    Instead of calling fwrite(), just call uv_default_loop() to get the
    jl_io_loop and then call jl_fs_write as usual.
    This means that all jl_printf()s now go through uvlib.
    There is a chance that this won’t work in win32, AppVeyor will tell...

Remove C interface for single character writes:

 - jl_fs_write_byte()
 - jl_putc_copy()
 - jl_pututf8_copy()

Hadnle single character writes in Julia:

 - write(f::File,        c::UInt8) = write(f, [c])
 - write(s::AsyncStream, b::UInt8) = write(s, [b])
 - write(s::AsyncStream, c::Char)  = write(s, string(c))

Rationale:

 - Writing to a libuv stream involves: "Julia does C malloc, Julia calls C,
   C calls libuv, libuv calls C callback, C callback calls julia hook,
   C callcback does c_free". Having less special cases of this makes
   it easier to reason about correctness.

 - Whatever overhead there might be in creating the temporary 1
   element array (or string); it is small compared to all the
   machinery in libuv, and the OS that actually makes the single
   character write happen.

 - If a user is doing single character writes to a libuv stream,
   they either don’t care about performance, or they should be
   buffering in Julia before calling write().

Inline jl_write_copy() into only remaining called: jl_write():

 - This is now the only place that needs to know about the
   trick used to store a copy of the output string after the
   request struct (i.e. malloc(sizeof(uv_write_t) + n));

Removed dead code:
 - stream.jl : make_stdout_stream()
 - stream.jl : jl_write_copy()

Rename:
 - jl_write_no_copy() -> jl_uv_write()

Move common code from stream.jl write() functions into uv_write macro:
 - ccall to jl_uv_write() (was jl_write_no_copy).
 - wait for current task

Simplify uv_write memory management:

Try to keep all the state management in one place (in macto uv_write()).

Move free() of uv_write_t request object
from C callback jl_uv_writecb_task() to macro uv_write().
This puts the malloc() and the free() in the same place.

Move uv_req_set_data(uvw,C_NULL) from C jl_uv_write() to macro
uv_write().
In _uv_hook_writecb_task() assert that uv_req_data(req) is not NULL.
libuv should not call the callback from within the call to uv_write(),
libuv has an internal “write_completed_queue” that is pumped by the main
loop.

samoconnor commented:
rebased and squashed, waiting for CI results...

tkelman commented:
AppVeyor looked like it was timing out on Win64 but it may have been unrelated. I queued up another build, if it happens again then maybe it's not a coincidence.

samoconnor commented:
AppV' and Travis both happy now.

tkelman commented:
Yeah, looks encouraging. @vtjnash do you have time to look over this one last time and/or merge before it goes conflict-stale again?

ivarne commented:
Bump!

vtjnash commented on src/codegen.cpp:
```diff
@@ -5026,9 +5016,9 @@ extern "C" void jl_init_codegen(void)
+    //jl_printf(JL_STDERR,"%s\n",jl_ExecutionEngine->getDataLayout()->getStringRepresentation().c_str());
     if (!jl_ExecutionEngine) {
-        JL_PRINTF(JL_STDERR, "Critical error initializing llvm: ", ErrorStr.c_str());
+        jl_printf(JL_STDERR, "Critical error initializing llvm: ", ErrorStr.c_str());
```
jl_errorf ?

samoconnor replied:
That would probably have the same effect.
However, per my note Jan 19, the time I can spend on this PR is now very limited.
I'll attempt to make things right if places are found where I've broken something.
But, I dont intend to add any more new changes on top of the existing refactoring.

vtjnash commented on src/task.c:
```diff
@@ -728,16 +728,16 @@ void NORETURN throw_internal(jl_value_t *e)
         assert(0);
     }
-    jl_exit(1);
+    assert(0);
```
i don't think compilers recognize the non-DEBUG version of `assert` as NORETURN, so this change may cause a warning when building the release version

samoconnor replied:
finish_task() and jl_longjmp() are both NORETURN, so the compiler should see that both assert(0)s are unreachable.
assert(0) is a better way to tell the human reader of the code that "this line is not reached".
If I was cleaning up this function a bit more I'd probably remove flatten out the else clause and put another assert(0) after jl_longjump().

BTW, I did not realise that Julia disables assert() in release builds. My preference is to always ship with assert() enabled (I usually implement some kind of user-friendly assert failure message / logging system).

vtjnash replied:
ah, right. i forgot jl_longjmp was also marked NORETURN

that's more-or-less just the default behavior for C compilers. there is a small performance hit from having them on everywhere, although I do tend to try to use them (esp. w/ setting `LLVM_ASSERTIONS=1`)

vtjnash commented on ui/repl.c:
```diff
@@ -139,8 +139,8 @@ void parse_opts(int *argcp, char ***argvp)
         case 'h':
-            ios_printf(ios_stdout, "%s%s", usage, opts);
-            exit(0);
+            jl_printf(JL_STDERR, "%s%s", usage, opts);
```
i think help messages usually go to `stdout`, not `stderr`

samoconnor replied:
fixed

vtjnash replied:
thanks

vtjnash commented on src/task.c:
```diff
@@ -696,11 +696,11 @@ DLLEXPORT void gdblookup(ptrint_t ip)
+            jl_safe_printf("unknown function (ip: %d)\n", line_num);
         else if (line_num == -1)
-            ios_printf(ios_stderr, "%s at %s (unknown line)\n", func_name, file_name, line_num);
+            jl_safe_printf("%s at %s (unknown line)\n", func_name, file_name, line_num);
```
(not your fault) but this function has the wrong number of arguments

samoconnor replied:
fixed

vtjnash replied:
thanks

vtjnash commented on src/profile.c:
```diff
@@ -216,7 +218,7 @@ void *mach_profile_listener(void *arg)
             }
             else if (forceDwarf == -1) {
-                JL_PRINTF(JL_STDERR, "Warning: Profiler attempt to access an invalid memory location\n");
+                jl_printf(JL_STDERR, "Warning: Profiler attempt to access an invalid memory location\n");
```
it's also illegal to use jl_* functions from this thread

samoconnor replied:
changed to jl_safe_printf()

vtjnash commented on src/jl_uv.c:
```diff
@@ -681,11 +602,20 @@ int jl_printf(uv_stream_t *s, const char *format, ...)
+    vsnprintf(buf, sizeof(buf), fmt, args);
+    va_end(args);
+
+    write(STDERR_FILENO, buf, strlen(buf));
```
buf may not be null-terminated. instead use the smaller of `sizeof(buf)` and `vsnprintf` return value (if > 0)

samoconnor replied:
man vsnprintf says "The snprintf() and vsnprintf() functions will write at most n-1 of the characters printed into the output string [...] The output is always null-terminated."

vtjnash replied:
i don't see any such statement: http://linux.die.net/man/3/vsnprintf. it seems that guarantee may be a BSD feature.

microsoft explicitly doesn't include it: https://msdn.microsoft.com/en-us/library/1kt27hek.aspx

samoconnor replied:
linux man per link above says "The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str."
i.e. the '\0' is included is the written bytes.

POISX says "a null byte is written at the end of the bytes actually written into the array." -- http://pubs.opengroup.org/onlinepubs/9699919799/functions/fprintf.html#

MS do their own thing. #9624 is related.
Perhaps it's better to have a compliant wrapper for vsnprintf() on win32 rather than make every call site aware of all the MS weirdness ?

vtjnash commented on src/codegen.cpp:
```diff
@@ -4748,15 +4747,6 @@ static void init_julia_llvm_env(Module *m)
-                         Function::ExternalLinkage,
-                         "jl_puts", m);
-    add_named_global(jlputs_func, (void*)&jl_puts);
-
```
the `jlstderr_var` could go too. i haven't needed to JIT a call to `jl_puts(jlstderr_var)` in quite some time. the addition of `code_llvm` was probably what mostly replaced that need.

vtjnash commented on base/stream.jl:
```diff
@@ -726,93 +725,54 @@ end
+        uv_req_set_data(uvw,ct)
+        ct.state = :waiting
+        stream_wait(ct)
+        c_free(uvw)
```
this needs a `try/finally` since you moved the `c_free` here

vtjnash replied:
although this seems to be less InterruptException-safe than it was before. perhaps also need to move the JL_SIGATOMIC into this function, since the code it was protecting also moved here

samoconnor replied:
There is still JL_SIGATOMIC in C code (in jl_uv_write()) around the call to uv_write().
Which code do you think could be unsafe? uv_req_set_data(uvw, C_NULL)?

vtjnash replied:
mostly the possible `c_free` interactions here

vtjnash commented on base/stream.jl:
```diff
@@ -726,93 +725,54 @@ end
 #end
 
-macro uv_write(n,call)
+macro uv_write(stream, data, len)
```
this should be a function

samoconnor replied:
done

samoconnor committed:
Refinement per @vtjnash comments.

Make uv_write() a function (was a macro) and do c_free in "finally" clause.

Explicit null-termination in jl_safe_printf() to allow for win32's
non-posix vsnprintf behaviour. See also #9624.

Use jl_safe_printf() in profile thread.

Send help message to jl_printf(JL_STDOUT,)  was stderr.

samoconnor commented:
bump - see comment Jan 19

tkelman commented:
Which one? It went conflict-stale again like I thought it would. Will you have time to rebase this, or will @vtjnash have to adopt the branch to get it rebased for merging? He's the main person qualified to make the final decision on it.

samoconnor commented:
rebased again...

samoconnor commented:
rebased again for review by @vtjnash

samoconnor commented:
rebased again for review by @vtjnash (following jl_errorf() change by @garrison https://github.com/JuliaLang/julia/commit/b66e5c10db88f1e8055be6a370e1f91e745f5bd9)

JeffBezanson commented on base/fs.jl:
```diff
@@ -177,14 +177,7 @@ function write(f::File, buf::Ptr{UInt8}, len::Integer, offset::Integer=-1)
-    uv_error("write",err)
-    1
-end
+write(f::File, c::UInt8) = write(f,[c])
```
I would worry about the performance of allocating an array here.

JeffBezanson replied:
Could be handled by `jl_fs_write`, passing `&c` as the data.

samoconnor replied:
I would argue that this is not the right layer to deal with performance of single-byte writes.
The OS implementation of the single-byte write involves a sys-call, maybe a network stack, etc... it is expensive.
If the caller needs to efficiently write lots of single bytes, they should write to a buffer interface that wraps the raw io interface.
Is there a particular use-pattern in current Julia code where single character writes to File are performance critical?

vtjnash replied:
you have to be careful there, since we could execute a task switch and invalid the stack slot before it gets written out

samoconnor replied:
Hi @vtjnash,
Sorry, I'm afraid I don't quite understand.
Are you saying that the user of a buffered interface might write a character that does not get sent because it is left in the buffer? ... in that case they should call buffer.flush(). No?

vtjnash replied:
that was addressed to Jeff's comment. just bad timing as we both typed replies at the same time

samoconnor commented:
still conflict clean at this point...

JeffBezanson commented:
Thanks, very nice cleanup.

ihnorton commented:
:+1:
